Add double-factorial implementation

rgizz opened the pull request:
Resolves #44 .

## Description

> What is the purpose of this pull request?

This pull request:

-   adds an implementation of the double factorial function.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #44 
-   is related to prior attempts to add similar functionality: https://github.com/stdlib-js/stdlib/pull/72 and https://github.com/rreusser/stdlib/commit/255f520dc5a901cd2b6af132a445cdc0d34281b8.

## Questions

> Any questions for reviewers of this pull request?

- Should the implementation be extended to include all reals? Complex numbers?

  - Related discussion: https://github.com/stdlib-js/stdlib/pull/72#discussion_r92948961

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

Prior art:

- Boost: https://github.com/boostorg/math/blob/develop/include/boost/math/special_functions/factorials.hpp#L74
- SciPy: https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.factorial2.html and https://github.com/scipy/scipy/blob/v1.11.3/scipy/special/_basic.py#L2940-L3014

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

rgizz committed:
Add double-factorial implementation

rgizz committed:
Add double-factorial test implementation

rgizz committed:
Add README.md file

rgizz committed:
typo to README

rgizz committed:
typo to README

rgizz committed:
Style edits, cleanup, and added license to main.js

rgizz committed:
Added n=0 case; added tests; six tests working

rgizz committed:
Revised README and recast in original Markdown

kgryte commented:
As discussed with regard to extension to complex numbers (https://en.wikipedia.org/wiki/Double_factorial#Extensions), the gamma approximation does not agree for even integers. As such, I am inclined to think we should stick with the "classic" definition.

We could, in theory, extend to negative integers; however, I don't think there's an immediate use case, so leaving as restricted to nonnegative integers, seems fine with me.

kgryte commented:
Additionally, I think that manually re-computing via a loop is fine. A look-up table would likely be a bit faster; however, a static table would significantly increase bundle sizes for what is likely to be a function which is not heavily used.

rgizz committed:
test: add license to test.js

This commit adds license to test.js per project requirements.

rgizz committed:
style: edit code to conform to style guide; return NaN for bad input

This commit (1) edits code to conform to style guide; changes 'let'
to 'var', and returns NaN for bad input.

rgizz committed:
feat: add benchmark.js file

rgizz committed:
feat: minor edits to main.js

This commit fixes typo in variable name and edits to conform to style guide.

rgizz committed:
test: add unit test

rgizz committed:
bench: delete unused code (decimals)

This commit deletes unused decimals code in benchmark.

rgizz committed:
revert: add back package.json (inadvertently omitted)

As stated in short message.

rgizz committed:
chore: add repl.txt file

This commit adds in repl.txt, a standard file not included when
project was started.

rgizz committed:
chore: moving README to proper place

As stated above.

rgizz committed:
chore: add .gitignore

To stop git from listing files as untraced in ../factorial directory.

rgizz committed:
chore: adding .ts files as per standard practice.

no-verify commit per A. Reines as typescript linter is to be
replaced.

rgizz committed:
chore: add README.md

This commit moves README.md from the wrong directory and
corrects some lint errors.

rgizz committed:
chore: restore index.js

Restore after inadvertently deleting.

rgizz committed:
chore: add integers.json

This commit adds integers.json, used for testing factorial2. NB these
values are not from Julia but hand calculated.

rgizz committed:
chore: add package.json to test directory

This file is copied and adapted from factorial directory.

rgizz committed:
chore: add index.js to examples directory

As stated in commit header.

rgizz committed:
chore: remove package.json

This commit corrects an error as said file should not have
been in this directory in the first place.

rgizz committed:
chore: remove .gitignore which was unnecessary

Had added .gitignore but not needed all derived from inadvertent
copying of files into wrong directory.

kgryte committed:
chore: update copyright year

kgryte committed:
bench: simplify benchmark to prioritize function execution over number generation

kgryte committed:
docs: fix examples and remove decimals

Decimals are removed as the input values are always expected to be
integers, and, similarly, the return values for input values within
range are also expected to be integers.

kgryte committed:
style: remove decimal

kgryte committed:
docs: update descriptions and fix duplicate variable declarations

kgryte committed:
docs: update example

kgryte committed:
chore: remove stray file

kgryte committed:
docs: fix examples

kgryte committed:
docs: rename variable

kgryte committed:
refactor: avoid nested fall-through and add asm type annotations

kgryte committed:
refactor: add asm type annotations

kgryte committed:
refactor: reorganize code and assign input value to typed internal variable

kgryte committed:
chore: update keywords and description

kgryte committed:
docs: insert HTML markup and update description and link

kgryte committed:
style: remove forced linebreaks

kgryte committed:
test: move and update fixtures to cover more cases and update tests

kgryte committed:
bench: add SciPy benchmarks

kgryte committed:
bench: enforce integer arithmetic

kgryte committed:
bench: add Boost benchmarks

kgryte committed:
style: remove extra space

kgryte committed:
style: remove extra space

kgryte committed:
style: remove trailing spaces

kgryte committed:
test: fix description

kgryte reviewed (approved):
LGTM. Thanks, @rgizz!
